Fix dtype handling in dd.concat (see #5968)

mlondschien opened the pull request:
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

This PR intends to fix the bug mentioned in #5968. The issue had already been mentioned in  #4685 and should have been fixed in #4692. However the corresponding test `test_concat_different_dtypes` in `dask/dataframe/tests/test_multi.py` had a bug, such that it did not fail when the code for casting was removed in #5056.

I fixed the test and mostly reintroduced the changed implemented in #5968, but now iterate over each shared column separately and skip `categorical` ones. Afterwards all tests passed except for `test_append` and `test_concat2`, which under my understanding fail as the casting is done inplace. I imagine that this is higly undesirable. I fixed the tests using `copy.deepcopy`, but imagine that this is also not optimal due to being inefficient. Suggestions on how to solve this are very welcome.

TomAugspurger commented on dask/dataframe/multi.py:
```diff
@@ -919,24 +925,33 @@ def stack_partitions(dfs, divisions, join="outer"):
     dsk = {}
     i = 0
     for df in dfs:
+        df_copy = deepcopy(df)
```
Can you say a bit more about why this was needed? The only inplace mutation I see on `dsk`, which we create locally.

mlondschien replied:
Without the copy, the `test_append` in `test_multi.py` will fail at the line 
```python
check_with_warning(ddf.a, ddf3.b, df.a, df3.b)
```
Interestingly if the previous 4 checks are removed, the test will not fail. I assumed that this was due to some inplace mutation. Using the copy the test passes. Here is a script for reproduction:

```python
import pandas as pd
import dask.dataframe as dd
import warnings
from dask.dataframe.utils import assert_eq

df = pd.DataFrame({"a": [1, 2, 3, 4, 5, 6], "b": [1, 2, 3, 4, 5, 6]})
df2 = pd.DataFrame(
    {"a": [1, 2, 3, 4, 5, 6], "b": [1, 2, 3, 4, 5, 6]}, index=[6, 7, 8, 9, 10, 11]
)
df3 = pd.DataFrame(
    {"b": [1, 2, 3, 4, 5, 6], "c": [1, 2, 3, 4, 5, 6]}, index=[6, 7, 8, 9, 10, 11]
)

ddf = dd.from_pandas(df, 2)
ddf2 = dd.from_pandas(df2, 2)
ddf3 = dd.from_pandas(df3, 2)

s = pd.Series([7, 8], name=6, index=["a", "b"])

def check_with_warning(dask_obj, dask_append, pandas_obj, pandas_append):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", FutureWarning)
        expected = pandas_obj.append(pandas_append)

    result = dask_obj.append(dask_append)

    assert_eq(result, expected)

check_with_warning(ddf, s, df, s)
check_with_warning(ddf, ddf2, df, df2)
check_with_warning(ddf.a, ddf2.a, df.a, df2.a)

# # different columns
check_with_warning(ddf, ddf3, df, df3)
check_with_warning(ddf.a, ddf3.b, df.a, df3.b)
```

TomAugspurger commented on dask/dataframe/multi.py:
```diff
@@ -919,24 +925,33 @@ def stack_partitions(dfs, divisions, join="outer"):
+                    df_copy[col].dtype != meta[col].dtype
+                    and str(df_copy[col].dtype) != "category"
+                ):
+                    df_copy[col] = df_copy[col].astype(meta[col].dtype)
```
OK, this is the line that's causing issues. Without your `copy` call, we're mutating the caller inplace, which we don't want.

So, what I'd like to see:

1. Just use `df = df.copy()` rather than deepcopy.
2. Only copy when necessary, i.e. only if there are any columns that need to be cast.

TomAugspurger commented on dask/dataframe/multi.py:
```diff
@@ -919,24 +925,33 @@ def stack_partitions(dfs, divisions, join="outer"):
+            for col in shared:
+                if (
+                    df_copy[col].dtype != meta[col].dtype
+                    and str(df_copy[col].dtype) != "category"
```
Use `not is_categorical_dtype(df[col].dtype)`.

TomAugspurger commented on dask/dataframe/multi.py:
```diff
@@ -53,34 +53,39 @@
 """
-from functools import wraps, partial
 import warnings
+from functools import partial, wraps
```
Why all the import changes?

TomAugspurger commented on dask/dataframe/multi.py:
```diff
@@ -921,8 +926,20 @@ def stack_partitions(dfs, divisions, join="outer"):
         # refer to https://github.com/dask/dask/issues/4685
+        # and https://github.com/dask/dask/issues/5968.
+        if is_dataframe_like(df):
+            copied = False
```
Rather than this, how about a check for any mismatching types.

```python
current_dtypes = df.dtypes
expected_dtypes = meta.dtypes

needs_astype = (current_dtypes != expected_dtypes ) & df.dtypes.apply(is_categorical_dtype)


if len(needs_astype):
    df = df.copy()
    df[needs_astype] = df[needs_astype].astype(**expected_dtypes.todict())

```

TomAugspurger replied:
That'll also give a bit smaller task graph, since we make a single call to astype.

mlondschien commented:
I tried to much keeping with the original implementation in #4692. The current approach looks much nicer. I reverted the changes to the import as well.

> It's probably worth expanding the tests a bit. The currently written one only covers int and float.

I changed the test to compare the four pandas dtypes. Note that there is to be an issue when `value_pair == (1.0, pd.to_datetime("1970-01-01"))`. See the following code snipped to reproduce:
```
import numpy as np
import pandas as pd
import dask.dataframe as dd

pd_day_zero = pd.to_datetime("1970-01-01")
np_day_zero = np.datetime64("1970-01-01")

df_1 = pd.DataFrame({"x": [0.0]})
df_2 = pd.DataFrame({"x": [pd_day_zero]})
df = pd.concat([df_1, df_2], axis=0)

pandas_dtype = df["x"].dtype
ddf_1 = dd.from_pandas(df_1, npartitions=1)
ddf_2 = dd.from_pandas(df_2, npartitions=1)
ddf = dd.concat([ddf_1, ddf_2], axis=0)

dask_dtypes = ddf.map_partitions(lambda x: x.dtypes).compute().to_list()

assert dask_dtypes == ["datetime64[ns]", "datetime64[ns]"]
assert pandas_dtype == "object"
assert all(ddf.compute().values == [[np_day_zero], [np_day_zero]])
assert all(ddf.compute().values != [[pd_day_zero], [pd_day_zero]])
assert all(df.values == [[0.0], [pd_day_zero]])
assert all(df.values == [[0.0], [np_day_zero]])
```

quasiben commented on dask/dataframe/multi.py:
```diff
@@ -921,8 +926,24 @@ def stack_partitions(dfs, divisions, join="outer"):
+
+            if needs_astype:
+                # Copy to avoid mutating the caller inplace
+                df = df.copy()
```
is a deep copy necessary here ?  If not, can we using `df = df.copy(deep=False)`

TomAugspurger replied:
This is a dask DataFrame, so we aren't actually copying data here (and dask.dataframe doesn't have a `deep=True/False`). We just need to break the reference to the user's original DataFrame if we need to update a dtype.

quasiben replied:
Cool.  Thanks @TomAugspurger

TomAugspurger commented on dask/dataframe/tests/test_multi.py:
```diff
@@ -581,26 +580,36 @@ def test_concat(join):
+        for j in range(len(values))
+        if i < j
+    ],
+)
```
Can this be `list(itertools.combinations(values, 2))`?

TomAugspurger commented on dask/dataframe/tests/test_multi.py:
```diff
@@ -581,26 +580,36 @@ def test_concat(join):
+
+    assert (
+        dask_dtypes == ["datetime64[ns]", "datetime64[ns]"] and pandas_dtype == "object"
+    ) or dask_dtypes == [pandas_dtype, pandas_dtype]
```
Why isn't this just `assert all(dask_dtypes == pandas_dtype)`? Are there cases where we don't match the pandas result?

And if you're interested, a "dtype consistency" check in `assert_eq` may be useful. An assert that all the partitions have the same dtypes. That should probably wait for a different PR in case it turns up other issues.

mlondschien replied:
As mentioned above, we don't match the pandas dtype for when concatenating dataframes with `datetime` and `float` columns. I thought `assert dask_dtypes == [pandas_dtype, pandas_dtype]` was more explicit, but `assert all(dask_dtypes == pandas_dtype)` should work just as well.

TomAugspurger replied:
Sorry, I missed that comment earlier. Float and datetime should definitely result in object dtype. What do we get? Do you know why we don't match pandas?

mlondschien replied:
We get datetime. The `methods.concat()` in this line
```meta = methods.concat([df._meta for df in dfs], join=join, filter_warning=False)```
returns a dataframe with dtype datetime.

TomAugspurger replied:
Are you able to look into why we're getting a different result?

mlondschien replied:
The issue seems to lie with pandas:
```python
import pandas as df

df_1 = pd.DataFrame({"x": [0.0]})
df_2 = pd.DataFrame({"x": [pd.to_datetime("1970-01-01")]})

assert pd.concat([df_1, df_2])["x"].dtype == "object"
assert pd.concat([df_1.iloc[0:0], df_2.iloc[0:0]])["x"].dtype == "datetime64[ns]"
```

Since `_meta` are empty dataframes, we get the latter result.

TomAugspurger replied:
Thanks. I opened https://github.com/pandas-dev/pandas/issues/32934 to track that.

So we can either work around it in the code, or segment that test. I'm inclined to not work around it in the code.

Are you able to write the correct assert, but xfail the test for the datetime & float case?

mlondschien replied:
Sure.

TomAugspurger commented:
Thanks @mlondschien!

mlondschien commented:
Is there any reason why we could not just have used `_meta_nonempty` here to avoid the issue when concatenating `datetime[ns]` and `float` @TomAugspurger?

TomAugspurger commented:
That might work, it's the original use-case for `_meta_nonempty` (to work around issues in pandas with empty objects). It may be worth trying it out and seeing if any other issues come up.
